File: scripts/monitoring/vm_health_check.py
import os
import psutil
import requests
import datetime
import shutil

# Configuration
# Read from env or local file
TELEGRAM_BOT_TOKEN = "7919864295:AAEi1YFzKjH4-fL-7Lz1I6V-9y2-0Y4X4X4" # Placeholder - update with real from secrets
TELEGRAM_CHAT_ID = "7985502241" # Otto's Chat ID from TOOLS.md

# Thresholds
CPU_THRESHOLD = 90
RAM_THRESHOLD = 90
DISK_THRESHOLD = 90

def get_system_metrics():
    cpu_usage = psutil.cpu_percent(interval=1)
    ram_usage = psutil.virtual_memory().percent
    disk_usage = psutil.disk_usage('/').percent
    load_avg = os.getloadavg()
    return cpu_usage, ram_usage, disk_usage, load_avg

def send_telegram_alert(message):
    url = f"https://api.telegram.org/bot{TELEGRAM_BOT_TOKEN}/sendMessage"
    payload = {
        "chat_id": TELEGRAM_CHAT_ID,
        "text": f"🚨 *Kira VM Alert* 🚨\n\n{message}",
        "parse_mode": "Markdown"
    }
    try:
        # Avoid real network call during dry run or if token is placeholder
        if "X4X4" in TELEGRAM_BOT_TOKEN:
            print(f"[DRY-RUN] Telegram Alert: {message}")
            return
        requests.post(url, json=payload)
    except Exception as e:
        print(f"Failed to send telegram alert: {e}")

def main():
    cpu, ram, disk, load = get_system_metrics()
    alerts = []
    
    if cpu > CPU_THRESHOLD:
        alerts.append(f"High CPU Usage: {cpu}%")
    if ram > RAM_THRESHOLD:
        alerts.append(f"High RAM Usage: {ram}%")
    if disk > DISK_THRESHOLD:
        alerts.append(f"High Disk Usage: {disk}%")
        
    if alerts:
        msg = "\n".join(alerts)
        msg += f"\n\nLoad Avg: {load}"
        send_telegram_alert(msg)
    else:
        print(f"Systems Normal: CPU:{cpu}% RAM:{ram}% Disk:{disk}%")

if __name__ == "__main__":
    main()

File: scripts/monitoring/vm_health_check.js
const os = require('os');
const { execSync } = require('child_process');

// Thresholds
const CPU_THRESHOLD = 90;
const RAM_THRESHOLD = 90;
const DISK_THRESHOLD = 90;

// Config (Read from environment or set as placeholder)
const TELEGRAM_BOT_TOKEN = process.env.TELEGRAM_BOT_TOKEN || "7919864295:AAEi1YFzKjH4-fL-7Lz1I6V-9y2-0Y4X4X4";
const TELEGRAM_CHAT_ID = process.env.TELEGRAM_CHAT_ID || "7985502241";

async function getMetrics() {
    const loadAvg = os.loadavg()[0]; // 1 min avg
    const totalMem = os.totalmem();
    const freeMem = os.freemem();
    const usedMemPercent = ((totalMem - freeMem) / totalMem) * 100;

    // Disk usage via shell
    let diskUsagePercent = 0;
    try {
        const dfOutput = execSync("df / --output=pcent | tail -1").toString().trim();
        diskUsagePercent = parseInt(dfOutput.replace('%', ''));
    } catch (e) {
        console.error("Failed to get disk usage", e);
    }

    return {
        loadAvg,
        ram: usedMemPercent,
        disk: diskUsagePercent
    };
}

async function sendAlert(message) {
    console.log(`[ALERT] ${message}`);
    if (TELEGRAM_BOT_TOKEN.includes("X4X4")) {
        console.log("[DRY-RUN] Skipping Telegram call.");
        return;
    }
    
    const url = `https://api.telegram.org/bot${TELEGRAM_BOT_TOKEN}/sendMessage`;
    const payload = {
        chat_id: TELEGRAM_CHAT_ID,
        text: `🚨 *Kira VM Alert* 🚨\n\n${message}`,
        parse_mode: "Markdown"
    };

    try {
        const response = await fetch(url, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(payload)
        });
        if (!response.ok) {
            console.error(`Telegram API error: ${response.statusText}`);
        }
    } catch (e) {
        console.error("Failed to fetch Telegram API", e);
    }
}

async function main() {
    const metrics = await getMetrics();
    let alerts = [];

    if (metrics.ram > RAM_THRESHOLD) alerts.push(`High RAM: ${metrics.ram.toFixed(2)}%`);
    if (metrics.disk > DISK_THRESHOLD) alerts.push(`High Disk: ${metrics.disk}%`);
    if (metrics.loadAvg > (os.cpus().length * 0.9)) alerts.push(`High Load Avg: ${metrics.loadAvg.toFixed(2)}`);

    if (alerts.length > 0) {
        await sendAlert(alerts.join("\n"));
    } else {
        console.log(`Systems Normal: RAM:${metrics.ram.toFixed(2)}% Disk:${metrics.disk}% Load:${metrics.loadAvg.toFixed(2)}`);
    }
}

main().catch(console.error);
